src/mem3/src/mem3_reshard.erl

```diff
@@ -0,0 +1,919 @@
+    {noreply, State}.
+
+
+handle_cast({db_deleted, DbName}, State) ->
+    % Remove only completed jobs. Other running states would `fail` but
+    % job result would stick around so users can inspect them.
+    JobIds = jobs_by_db_and_state(DbName, completed),
+    [remove_job_int(JobId, State) || JobId <- JobIds],
+    {noreply, State};
+
+handle_cast({report, Job}, State) ->
+    report_int(Job),
+    {noreply, State};
+
+handle_cast({checkpoint, Job}, State) ->
+    {noreply, checkpoint_int(Job, State)};
+
+handle_cast(reload_jobs, State) ->
+    couch_log:notice("~p starting reloading jobs", [?MODULE]),
+    State1 = reload_jobs(State),
+    couch_log:notice("~p finished reloading jobs", [?MODULE]),
+    {noreply, State1};
+
+handle_cast(Cast, State) ->
+    couch_log:error("~p unexpected cast ~p", [?MODULE, Cast]),
+    {noreply, State}.
+
+
+handle_info({'DOWN', _Ref, process, Pid, Info}, State) ->
+    case job_by_pid(Pid) of
+        {ok, Job} ->
+            couch_log:notice("~p job ~s exit ~p", [?MODULE, Job#job.id, Info]),
+            ok = handle_job_exit(Job, Info, State);
+        {error, not_found} ->
+            couch_log:error("~p job not found: ~p ~p", [?MODULE, Pid, Info])
+    end,
+    {noreply, State};
+
+handle_info(Info, State) ->
+    couch_log:error("~p unexpected info ~p", [?MODULE, Info]),
+    {noreply, State}.
+
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+
+%% Private API
+
+validate_and_start_job(#shard{} = Source, Split) ->
+    case mem3_reshard_validate:start_args(Source, Split) of
+        ok ->
+            Target = target_shards(Source, Split),
+            case mem3_reshard_validate:targets(Source, Target) of
+                ok ->
+                    TStamp = now_sec(),
+                    Job = #job{
+                        type = split,
+                        job_state = new,
+                        split_state = new,
+                        start_time = TStamp,
+                        update_time = TStamp,
+                        node = node(),
+                        source = Source,
+                        target = Target
+                    },
+                    Job1 = Job#job{id = job_id(Job)},
+                    Job2 = update_job_history(Job1),
+                    gen_server:call(?MODULE, {start_job, Job2}, infinity);
+                {error, Error} ->
+                    {error, Error}
+            end;
+        {error, Error} ->
+            {error, Error}
+    end.
+
+
+handle_start_job(#job{} = Job, #state{state = running} = State) ->
+    case start_job_int(Job, State) of
+        ok ->
+            {reply, {ok, Job#job.id}, State};
+        {error, Error} ->
+            {reply, {error, Error}, State}
+    end;
+
+handle_start_job(#job{} = Job, #state{state = stopped} = State) ->
+    ok = mem3_reshard_store:store_job(State, Job),
+    % Since resharding is stop cluster-wide, job is
```

Comment:
I'll have to re-word that, thanks for pointing it out!

What I meant to say is if the resharding is stopped globally (user PUTs "state":"stopped" to /_reshard/state), when examining the jobs list it would be odd to see jobs being an the `running` state. So we make sure to `temporarily` stop it  which forces the state in ets to be marked as `stopped`. But that's temporary, and we don't persist it. The store (the _local) doc will keep the job marked `running` so that when the global state toggle to running the job will be resumed and running.